### PR TITLE
quickTunerPreproc.py, preprocessor script for quick tuner scripts

### PR DESCRIPTION
**Pull request by ethansaurusrex**
This script accepts the *.debug files that are generated from tuningRunner.py as input and concatenates them into a single tsv file for use in the next step of the quick tuning pipeline: generation.

### CODE REVIEW

File: mlir/utils/performance/quickTunerPreproc.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+class qtPreprocessor(object):
+    """
+    class to process *.debug files into a single script
+    """
+
+    def __init__(self, pargs):
+        self.input_dir = pargs.input_dir
+
+    @staticmethod
```
Comment (pcf000) [inline mlir/utils/performance/quickTunerPreproc.py:56]: Why staticmethod?  Not a criticism, I'm just less familiar with why that's done in python and I'm curious.

File: mlir/utils/performance/quickTunerPreproc.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        tsv_files = glob.glob(os.path.join(input_dir, f"*.{file_ext}"))
+        print(os.path.join(input_dir, f"*.{file_ext}"))
+
+        dfs = []
+        ct = 0
+        for file in tsv_files:
+            df = pd.read_csv(file, sep='\t')
+            if normalize:
+                scaler = MinMaxScaler()
+                df['TFlops'] = scaler.fit_transform(df[['TFlops']])
```
Comment (pcf000) [inline mlir/utils/performance/quickTunerPreproc.py]: Why one set of brackets on the left but two on the right??

File: mlir/utils/performance/quickTunerPreproc.py (hunk 1 above)
Comment (ethansaurusrex) [reply mlir/utils/performance/quickTunerPreproc.py:56]: This was intended to be imported and called within the quickTunerGen.py set of classes but I could not rationalize each individual class having their own instance of this preprocessor class hanging out inside:
So, it was either have a standalone function or a class with a static method that would allow me to do something like:
```
qtPreprocessor.process( /* args */ )
```
From my experience with using the `@staticmethod` decorator, it has always been what to use when you want to package functions together that have similar functionality and can share data but do not share state.

File: mlir/utils/performance/quickTunerPreproc.py (hunk 2 above)
Comment (ethansaurusrex) [reply mlir/utils/performance/quickTunerPreproc.py]: I think that was a mess up on my end. Fortunately (or unfortunately) the syntax is forgiving. 

Also, this will actually be changed to add a `NormalizedTFlops` column instead of overwriting the current one.
```
if normalize:
    scaler = MinMaxScaler()
    df['NormalizedTFlops'] = scaler.fit_transform(df[['TFlops']])
```

File: mlir/utils/performance/quickTunerPreproc.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    cmd = env_str + f" python3 ./bin/tuningRunner.py -q {special_args} \
+                     --config='{config_string}' --mlir-build-dir `pwd` \
+                     --output=- --tflops \
+                     --rocmlir_gen_flags='--device={self.gpu_id}' 2>/dev/null"
+
+
+to:
+    
+    import uuid
```
Comment (krzysz00) [inline mlir/utils/performance/quickTunerPreproc.py:25]: Could this instead be a patch to `rocmlir_worker`?

File: mlir/utils/performance/quickTunerPreproc.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+"""
+
+import os
+import sys
+import argparse
+import pandas as pd
+import glob
+from sklearn.preprocessing import MinMaxScaler
+
+class qtPreprocessor(object):
```
Comment (krzysz00) [inline mlir/utils/performance/quickTunerPreproc.py]: Nit class case, and I'm pretty sure Python 3 doesn't require the `(object)` anymore

File: mlir/utils/performance/quickTunerPreproc.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        # num unique gemms in file:
+        cols = ['TransA', 'TransB', 'G', 'M', 'N','K']
+        unique_gemms = df[cols].drop_duplicates()
+        
+        num_gemms = len(unique_gemms)
+        print(f"Number of unique Gemms: {num_gemms}")
+        for _,row in unique_gemms.iterrows():
+            tup = tuple(row)
+            print(f"{tup[0]},{tup[1]},{tup[2]},{tup[3]},{tup[4]},{tup[5]}")
```
Comment (krzysz00) [inline mlir/utils/performance/quickTunerPreproc.py]: This is, I'd think, print(`,`.join(row)`)

File: mlir/utils/performance/quickTunerPreproc.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        """
+        staticmethod process() function that compiles output files into a single dataframe and saves to tsv file
+        """
+
+        tsv_files = glob.glob(os.path.join(input_dir, f"*.{file_ext}"))
+        print(os.path.join(input_dir, f"*.{file_ext}"))
+
+        dfs = []
+        ct = 0
+        for file in tsv_files:
```
Comment (krzysz00) [inline mlir/utils/performance/quickTunerPreproc.py:118]: You can `emumerate()` instead of maintaining a `ct`

File: mlir/utils/performance/quickTunerPreproc.py (hunk 5 above)
Comment (krzysz00) [reply mlir/utils/performance/quickTunerPreproc.py]: ... This could also be `to_csv` on the dataframe with `skip_index=True` or some such thing

**Review by krzysz00 — COMMENTED** [review]
Python nitpicking

File: mlir/utils/performance/quickTunerPreproc.py (hunk 3 above)
Comment (pcf000) [reply mlir/utils/performance/quickTunerPreproc.py:25]: Not in its current form;  it cuts off the usual functionality of passing results to stdout and we don't have code to read results from the files and send it on.  I do have an open issue about collecting all the results into the database and this may become part of it.

File: mlir/utils/performance/quickTunerPreproc.py (hunk 5 above)
Comment (ethansaurusrex) [reply mlir/utils/performance/quickTunerPreproc.py]: > This is, I'd think, print(`,`.join(row)`)

If the row contains non string entries, then it will complain and require conversion outside of the `.join()` function.

File: mlir/utils/performance/quickTunerPreproc.py (hunk 6 above)
Comment (ethansaurusrex) [reply mlir/utils/performance/quickTunerPreproc.py:118]: I will replace this with a `len()` call since it is not needed within the loop only for generating stats.

File: mlir/utils/performance/quickTunerPreproc.py (hunk 5 above)
Comment (krzysz00) [reply mlir/utils/performance/quickTunerPreproc.py]: `','.join(map(str, tup))`

File: mlir/utils/performance/quickTunerPreproc.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+"""
+
+import os
+import sys
+import argparse
+import pandas as pd
+import glob
+from sklearn.preprocessing import MinMaxScaler
+
+class qtPreprocessor():
```
Comment (krzysz00) [inline mlir/utils/performance/quickTunerPreproc.py:48]: Standard complaining re style on class names

**Review by krzysz00 — APPROVED** [review]
Broadly approved
